src/backend/distributed/metadata/metadata_utility.c

```diff
@@ -893,11 +893,28 @@ GenerateAllShardStatisticsQueryForNode(WorkerNode *workerNode, List *citusTableI
 	Oid relationId = InvalidOid;
 	foreach_oid(relationId, citusTableIds)
 	{
-		List *shardIntervalsOnNode = ShardIntervalsOnWorkerGroup(workerNode, relationId);
-		char *shardStatisticsQuery =
-			GenerateShardStatisticsQueryForShardList(shardIntervalsOnNode,
-													 useShardMinMaxQuery);
-		appendStringInfoString(allShardStatisticsQuery, shardStatisticsQuery);
+		/*
+		 * Ensure the table still exists by trying to acquire a lock on it
+		 * If function returns NULL, it means the table has been concurrently dropped
+		 * hence we should skip
+		 */
+		Relation relation = try_relation_open(relationId, AccessShareLock);
+		if (relation != NULL)
+		{
+			List *shardIntervalsOnNode = ShardIntervalsOnWorkerGroup(workerNode,
+																	 relationId);
+			char *shardStatisticsQuery =
+				GenerateShardStatisticsQueryForShardList(shardIntervalsOnNode,
+														 useShardMinMaxQuery);
+			appendStringInfoString(allShardStatisticsQuery, shardStatisticsQuery);
+			relation_close(relation, AccessShareLock);
```

Comment:
Note that it is still possible to drop the table after this line but since we will send the command optionally, we will get a warning from workers saying that the relation(shard) doesn't exist on the workers.

Reply:
Can we keep the lock until after the function is finished? In that case it's not possible to drop until the function is done, but I'm not sure about the consequences of that.

Reply:
If you close the relation with `NoLock` then the lock will be held until the transaction finishes (either aborted or committed). One consequence would be holding the locks for a long time. If we had many tables it is possible that citus_shards would block the drop operations and it can possibly take a long time until we can acquire all the locks on all the tables. So I think we shouldn't hold the locks until the transaction is finished and I think it is fine to get the warnings from the workers